### PR TITLE
[DE509848] 500 internal server error for RefreshToken call during DB maintenance.

### PR DESCRIPTION
**Pull request by ysd24**
500 internal server error for RefreshToken call during DB maintainance.

### CODE REVIEW

**Comment by graju256** [comment]
@ysd24  @arindamhit this change is similar to that of https://github.com/CAAPIM/Android-MAS-SDK/pull/397, but for iOS.

Please ensure the configuration/SDK visible method names similar across platforms. Of course, the behaviour of the change should also be the same.

File: MASFoundation/Classes/MAS.h [hunk 1]
```diff
@@ -176,6 +176,13 @@
 + (BOOL)isSSLPinningEnabled;
 
 
++ (void)setDonotLogoutRefreshtokenServerMaintainance:(BOOL)enable;
```
Comment (graju256) [inline MASFoundation/Classes/MAS.h]: this name looks different from the Android SDK implementation

Comment (graju256) [reply MASFoundation/Classes/MAS.h]: @arindamhit @ysd24 please conclude on this common name.

File: MASFoundation/Classes/_private_/services/model/MASModelService.m [hunk 2]
```diff
@@ -1991,6 +1993,33 @@ - (void)loginAsRefreshTokenWithCompletion:(MASCompletionErrorBlock)completion
                                           //
                                           if (error)
                                           {
+                                              //
+                                              // DE509848 - During server maintenance timeframe,
+                                              // when token refresh calls returns 500, MAS library logout the user.
+                                              //
+                                              if ([MAS isDonotLogoutRefreshtokenServerMaintainance] && httpResponse.statusCode == 500 &&
```
Comment (graju256) [inline MASFoundation/Classes/_private_/services/model/MASModelService.m]: This condition seemed more strict than the android version. Please ensure both the implementation has the common logic.

NOTE: As of now, I've no idea which one is more opting to the customer's scenario.

Comment (ysd24) [reply MASFoundation/Classes/_private_/services/model/MASModelService.m]: Removed the check for matching the error description and title.

File: MASFoundation/Classes/_private_/services/model/MASModelService.m [hunk 3]
```diff
@@ -1991,6 +1993,30 @@ - (void)loginAsRefreshTokenWithCompletion:(MASCompletionErrorBlock)completion
...
+                                              // when token refresh calls returns 500, MAS library logout the user.
+                                              //
+                                              if ([MAS isDonotLogoutTokenRenewalOnServerErrors] && httpResponse.statusCode == 500)  {
+                                                  
+                                                  NSError *idTokenValidationError = nil;
+                                                  BOOL isIdTokenValid = [MASAccessService validateIdToken:[[MASAccessService sharedService] getAccessValueStringWithStorageKey:MASKeychainStorageKeyIdToken]
+                                                                                            magIdentifier:[[MASAccessService sharedService] getAccessValueStringWithStorageKey:MASKeychainStorageKeyMAGIdentifier]
+                                                                                                    error:&idTokenValidationError];
+                                                  
+                                                  NSString *refreshToken =
```
Comment (graju256) [inline MASFoundation/Classes/_private_/services/model/MASModelService.m:2007]: I think we've not been counting on the refresh-token. Please check: https://github.com/CAAPIM/Android-MAS-SDK/pull/397/files#diff-4b5920088a9b497400ba73a15a524ed770fa6749f764fd9a20d5a786060e5bc4R261-R267

File: MASFoundation/Classes/_private_/services/model/MASModelService.m [hunk 4]
```diff
@@ -1991,6 +1993,30 @@ - (void)loginAsRefreshTokenWithCompletion:(MASCompletionErrorBlock)completion
...
+                                                    [MASAccessService sharedService].currentAccessObj.refreshToken;
+                                                  if (refreshToken && isIdTokenValid && !idTokenValidationError)
+                                                  {
+                                                      if (blockCompletion) {
+                                                      
+                                                          blockCompletion(NO, error);
+                                                      }
+                                                  
+                                                      return;
+                                                  }
```
Comment (graju256) [inline MASFoundation/Classes/_private_/services/model/MASModelService.m]: [minor] please stick to the one style of the indentation. Remove unnecessary empty lines.

Comment (ysd24) [reply MASFoundation/Classes/_private_/services/model/MASModelService.m]: Done.

**Review by arindamhit — APPROVED** [review]
Looks fine to me

**Pull request merged**